File: Objetos_2_java/src/main/java/objetos2/java/reservasHotel/GestionReservas.java
package objetos2.java.reservasHotel;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.locks.ReentrantLock;

public class GestionReservas {

    private ArrayList<Habitacion> habitaciones = new ArrayList<Habitacion>();

    private ArrayList<Reserva> reservas = new ArrayList<Reserva>();

    public void agregarHabitacion(Habitacion habitacion) {
        if (!habitaciones.contains(habitacion)) {
            habitaciones.add(habitacion);
            System.out.println("Habitacion numero " + habitacion.numero + " agregada correctamente");
        }else{
            System.out.println("Ya existe esta habitacion");
        }
    }

    public void mostrarHabsDispo(){
        List<Habitacion> habitacionesDispo = new ArrayList<Habitacion>();

        // Añadir las habitaciones disponibles a la lista
        for (Habitacion habitacion : habitaciones) {
            if (!habitacion.ocupada){
                habitacionesDispo.add(habitacion);
            }
        }


        if (!habitacionesDispo.isEmpty()) {
            System.out.println("Habitaciones disponibles: ");

            for (Habitacion habitacion : habitacionesDispo) {
                habitacion.mostrarInfo();
            }
        }else{
            System.out.println("No hay habitaciones disponibles");
        }
    }


    public void reservarHabitacion(Cliente cliente, Habitacion habitacion) {
        if (!habitacion.ocupada && habitaciones.contains(habitacion)) {
            Reserva reserva = new Reserva(habitacion, cliente);
            reservas.add(reserva);
            habitacion.ocupada = true;
            System.out.println("Reserva de la habitacion con numero " + habitacion.numero + " realizada correctemante por el cliente " + cliente.nombre);
        }else{
            System.out.println("Habitacion no disponible para reservar");
        }
    }


    public void cancelarHabitacion(Cliente cliente, Habitacion habitacion) {
        ArrayList<Reserva> reservasCancelar = new ArrayList<Reserva>();

        for (Reserva reserva : reservas) {
            if (Objects.equals(reserva.cliente.nombre, cliente.nombre) && Objects.equals(reserva.habitacion.numero, habitacion.numero)){
                reservasCancelar.add(reserva);
            }
        }

        if (!reservasCancelar.isEmpty()) {
            reservasCancelar.removeAll(reservas);
            habitacion.ocupada = false;
            System.out.println("Reserva del cliente " + cliente.nombre + " de la habitacion " + habitacion.numero + " cancelada correctamente");
        }else{
            System.out.println("No existe ninguna reserva para el cliente " + cliente.nombre + " y numero de habitacion " + habitacion.numero);
        }

    }

    public void mostrarReservas(){
        if (!reservas.isEmpty()) {
            System.out.println("Listado de reservas: ");
            for (Reserva reserva : reservas) {
                System.out.println("- Num habitacion: " + reserva.habitacion.numero + ", Cliente: " + reserva.cliente.nombre + ", Fecha reserva: " + reserva.fecha);
            }
        }else{
            System.out.println("No hay reservas actualmente");
        }
    }
}

File: Objetos_2_Kotlin/src/main/kotlin/reservasHotel/GestionReservas.kt
package reservasHotel

class GestionReservas {

    private val habitaciones = mutableListOf<Habitacion>()
    private val reservas = mutableListOf<Reserva>()

    fun agregarHabitacion(habitacion: Habitacion) {
        if (!habitaciones.contains(habitacion)) {
            habitaciones.add(habitacion)
            println("Habitación número ${habitacion.numero} agregada correctamente.")
        }else{
            println("Ya existe esta habitación.")
        }
    }


    fun mostrarHabsDispo(){
        val habDispos = habitaciones.filter { !it.ocupada }

        if (habDispos.isNotEmpty()) {
            println("Habitaciones disponibles: ")
            habDispos.forEach { println(it)}
        }else{
            println("No hay habitaciones disponibles.")
        }
    }


    fun reservarHabitacion(cliente: Cliente, habitacion: Habitacion) {
        if (!habitacion.ocupada && habitaciones.contains(habitacion)) {
            val reserva = Reserva(habitacion, cliente)
            reservas.add(reserva)
            habitacion.ocupada = true
            println("Reserva de la habitación numero ${habitacion.numero} realizada correctamente por el cliente ${cliente.nombre}")
        }else{
            println("Habitación no disponible para reservar")
        }
    }


    fun cancelarHabitacion(cliente: Cliente, habitacion: Habitacion) {
        val reserva = reservas.find {it.cliente == cliente && it.habitacion == habitacion}

        if (reserva != null) {
            reservas.remove(reserva)
            habitacion.ocupada = false
            println("Reserva del cliente ${cliente.nombre} de la habitacion numero ${habitacion.numero} cancelada correctamente.")

        }else{
            println("No existe ninguna reserva para el cliente ${cliente.nombre} y numero de habitacion ${habitacion.numero}")
        }
    }

    fun mostrarReservas(){
        if (reservas.isNotEmpty()) {
            reservas.forEachIndexed { index, reserva ->
                println("Reserva ${index+1}.- Num habitacion: ${reserva.habitacion.numero}, Cliente: ${reserva.cliente.nombre}, Fecha reserva: ${reserva.fechaReserva}")
            }
        }else{
            println("No hay reservas actualmente.")
        }
    }
}